applepay/utils.py

```diff
@@ -66,3 +100,316 @@ def signing_time_is_valid(signature, current_time, threshold):
+        logger.warning("Leaf Certificate OID not found")
+        return False
+
+    if not intermediate_cert:
+        logger.warning("Intermediate certificate OID not found")
+        return False
+
+    root_der = open(payment.ROOT_CA_FILE, 'r').read()
+
+    # pass the der-encoded representation of the certs
+    if not valid_chain_of_trust(root_der, intermediate_cert.dump(), leaf_cert.dump()):
+        return False
+
+    # Build the signer information from the signer that matches the leaf cert
+    signer_info = get_first_from_iterable(
+        filter_func=lambda signer: signer['sid'].chosen['serial_number'].native == leaf_cert.serial_number,
+        iterable=signed_data['signer_infos']
+    )
+
+    if not signer_info:
+        logger.warning("No signature found for the leaf cert.")
+        return False
+
+    # Use the signed attrs to verify the data was signed within the threshold
+    # provided and is from who it says its from
+    signed_attrs = signer_info['signed_attrs']
+
+    # only check the signing time if a threshold was provided
+    if threshold:
+        singing_time_attr = get_first_from_iterable(
+            filter_func=lambda signed_attr: signed_attr['type'].dotted == payment.OID_SIGNING_TIME,
+            iterable=signed_attrs
+        )
+
+        if not valid_signing_time(singing_time_attr['values'][0].native, datetime.now(utc), threshold):
+            logger.warning("Signing time outside of threshold.")
+            return False
+
+    message_digest_attr = get_first_from_iterable(
+        filter_func=lambda signed_attr: signed_attr['type'].dotted == payment.OID_MESSAGE_DIGEST,
+        iterable=signed_attrs
+    )
+
+    if not message_digest_attr:
+        logger.warning("No message digest found for the leaf cert.")
+        return False
+
+    message_digest = message_digest_attr['values'][0].native
+
+    data = get_payment_data(token)
+
+    # build the hashfunc from the leaf_cert's defined algorithm
+    hashfunc = partial(get_hashfunc_by_name, leaf_cert.hash_algo)
+
+    hashed_data = hashfunc(data).digest()
+
+    if hashed_data != message_digest:
+        logger.warning("Message digest does not match provided data.")
+        return False
+
+    signed_attrs_ber = get_ber_encoded_signed_attributes(signed_attrs)
+    public_key_point = remove_ec_point_prefix(leaf_cert.public_key['public_key'].native)
+
+    if not public_key_point:
+        return False
+
+    sig_octets = signer_info['signature'].native  # the signature to verify
+    # The signature is der-encoded, so use sigdecode_der to decode
+    # it later on
+    sigdecode = sigdecode_der
+
+    vk = VerifyingKey.from_string(public_key_point, curve=curves.NIST256p, hashfunc=hashfunc)
+
+    # verify that the signature matches the signed data
+    try:
+        vk.verify(sig_octets, signed_attrs_ber, hashfunc=hashfunc, sigdecode=sigdecode)
+    except BadSignatureError:
+        logger.warning("Invalid signature.")
+        return False
```

Comment:
It does seem like there is a lot going on in this function that could be delegated to helper methods, but I'm not up for premature optimization if we're going to continue to work on this library...